src/Symfony/Component/Translation/Tests/TranslatorCacheTest.php

```diff
@@ -51,117 +52,121 @@ protected function deleteTmpDir()
+            $translator->addLoader($format, new ArrayLoader());
+            $translator->addResource($format, array($msgid => 'OK'), $locale);
+            $translator->trans($msgid);
+
+            // Try again and see we get a valid result whilst no loader can be used
+            $translator = new Translator($locale, null, $this->tmpDir);
+            $translator->addLoader($format, $this->createFailingLoader());
+            $translator->addResource($format, array($msgid => 'OK'), $locale);
+            $this->assertEquals('OK', $translator->trans($msgid), '-> caching does not work in '.($debug ? 'debug' : 'production'));
+        }
     }
 
-    public function testRefreshCacheWhenResourcesChange()
+    public function testCatalogueIsReloadedWhenResourcesAreNoLongerFresh()
     {
-        // prime the cache
-        $loader = $this->getMock('Symfony\Component\Translation\Loader\LoaderInterface');
-        $loader
-            ->method('load')
-            ->will($this->returnValue($this->getCatalogue('fr', array(
-                'foo' => 'foo A',
-            ))))
-        ;
+        /*
+         * The testThatACacheIsUsed() test showed that we don't need the loader as long as the cache
+         * is fresh.
+         *
+         * Now we add a Resource that is never fresh and make sure that the
+         * cache is discarded (the loader is called twice).
+         */
 
-        $translator = new Translator('fr', new MessageSelector(), $this->tmpDir, true);
-        $translator->setLocale('fr');
-        $translator->addLoader('loader', $loader);
-        $translator->addResource('loader', 'foo', 'fr');
+        $locale = 'whatever';
+        $format = 'what-you-like';
+        $msgid = 'test';
 
-        $this->assertEquals('foo A', $translator->trans('foo'));
+        $catalogue = new MessageCatalogue($locale, array());
+        $catalogue->addResource(new StaleResource()); // better use a helper class than a mock, because it gets serialized in the cache and re-loaded
 
-        // add a new resource to refresh the cache
+        /** @var LoaderInterface|\PHPUnit_Framework_MockObject_MockObject $loader */
         $loader = $this->getMock('Symfony\Component\Translation\Loader\LoaderInterface');
         $loader
+            ->expects($this->exactly(2))
             ->method('load')
-            ->will($this->returnValue($this->getCatalogue('fr', array(
-                'foo' => 'foo B',
-            ))))
+            ->will($this->returnValue($catalogue))
         ;
 
-        $translator = new Translator('fr', new MessageSelector(), $this->tmpDir, true);
-        $translator->setLocale('fr');
-        $translator->addLoader('loader', $loader);
-        $translator->addResource('loader', 'bar', 'fr');
-
-        $this->assertEquals('foo B', $translator->trans('foo'));
+        // 1st pass
+        $translator = new Translator($locale, null, $this->tmpDir, /* debug --> */ true);
```

Comment:
you can remove ```/* debug --> */``` and instead add a comment above why we made test only in debug mode